Modules/Registration/Common/include/itkMeanSquaresImageToImageMetric.hxx

```diff
@@ -164,16 +164,30 @@ MeanSquaresImageToImageMetric<TFixedImage, TMovingImage>::GetValueAndDerivativeT
     transform = this->m_Transform;
   }
 
-  // Jacobian should be evaluated at the unmapped (fixed image) point.
-  transform->ComputeJacobianWithRespectToParameters(fixedImagePoint, threadS.m_Jacobian);
-  for (unsigned int par = 0; par < this->m_NumberOfParameters; ++par)
+  if (this->m_TransformIsBSpline && this->m_UseCachingOfBSplineWeights)
+  { // using pre-computed weights and indexes to calculate only non zero elements of the derivative
+    for (unsigned int k = 0; k < this->m_NumBSplineWeights; ++k)
+    {
+      for (unsigned int dim = 0; dim < MovingImageDimension; ++dim)
+      {
+        int par = this->m_BSplineTransformIndicesArray[fixedImageSample][k] + this->m_BSplineParametersOffset[dim];
+        threadS.m_MSEDerivative[par] += 2.0 * diff * this->m_BSplineTransformWeightsArray[fixedImageSample][k] * movingImageGradientValue[dim];
```

Comment:
I changed line 174 to 
threadS.m_MSEDerivative[par] = 99;

and no tests failed. 


We need to have a tests that executes this code before it can be added to the toolkit.  Currently the code is never run by any tests.

Reply:
Test has been added to itkMeanSquaresImageMetricTest.cxx that checks consistency between derivative computed omitting versus using weights caching. Tested on my end, can confirm it fails using 99.0 * ... and passes using 2.0 * ...

Reply:
@ljm898 We are looking forward to seeing these changes :).

Reply:
Looks like the code didn't push properly. Working on it now.